pkg/cmd/roachtest/operations/changefeeds/utils.go

```diff
@@ -0,0 +1,385 @@
+		case <-time.After(pollForStateInterval):
+			// the current state of the job is obtained.
+			jobToAction.state, jobToAction.highWaterTimestamp, err = getJobIDStateAndHighWM(ctxTimeout, conn, jobToAction.jobID)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// checkAndCreateChangefeed checks if there are any existing changefeed jobs,
+// and if the createOnlyIfNoJobs flag is true, it returns early if any jobs exist.
+// If the number of changefeed jobs is less than the maximum allowed limit,
+// it randomly selects a database and table, then attempts to create a new changefeed job for that table.
+// The function returns true if it successfully creates or evaluates the need for a changefeed,
+// and false if any errors occur during the process.
+func checkAndCreateChangefeed(
+	ctx context.Context, o operation.Operation, conn *gosql.DB, createOnlyIfNoJobs bool,
+) error {
+	// Fetch all changefeed jobs that are running or paused.
+	allCFJobs, err := fetchAllChangefeedJobsWithStates(ctx, conn, jobs.StateRunning, jobs.StatePaused)
+	if err != nil {
+		return err
+	}
+	// If there are existing changefeeds and the function is instructed to
+	// create only if no jobs exist, return early.
+	if len(allCFJobs) > 0 && createOnlyIfNoJobs {
+		o.Status(fmt.Sprintf("No jobs created as %d active jobs already exists", len(allCFJobs)))
+		return nil
+	}
+
+	// Get the max allowed changefeeds from environment variables.
+	mcf, err := helpers.EnvOrDefaultInt(maxChangefeeds, defaultEnvValuesInt[maxChangefeeds])
+	if err != nil {
+		return err
+	}
+
+	// If the number of existing changefeeds is below the max limit, create a new one.
+	if len(allCFJobs) < mcf {
+		// Select a random database and table to create the changefeed for.
+		dbName := helpers.PickRandomDB(ctx, o, conn, helpers.SystemDBs)
+		tableName := helpers.PickRandomTable(ctx, o, conn, dbName)
+
+		// Attempt to create the changefeed.
+		err = createChangefeed(ctx, o, conn, allCFJobs, dbName, tableName)
+		if err != nil {
+			return err
+		}
+		// createdJob is the job that was just created. It is identified by querying the
+		// changefeed job with the latest created timestamp
+		var createdJob *jobDetails
+		statement := "SELECT job_id, status, COALESCE(high_water_timestamp, '0') " +
+			"FROM [SHOW CHANGEFEED JOBS] order by created desc limit 1;"
+		err = helpers.ExecuteQuery(ctx, func(rowValues []string) error {
+			highWaterTimestamp, err := hlc.ParseHLC(rowValues[2])
+			createdJob = &jobDetails{
+				jobID:              rowValues[0],
+				state:              jobs.State(rowValues[1]),
+				payload:            nil,
+				highWaterTimestamp: highWaterTimestamp,
+			}
+			return err
+		}, conn, statement)
+		if err != nil {
+			return err
+		}
+		return waitForJobToComplete(ctx, conn, "CREATE", createdJob,
+			[]jobs.State{jobs.StateRunning, jobs.StateSucceeded})
+	} else {
+		o.Status(fmt.Sprintf("total changefeeds %d are more than or equal to the maximum changefeeds criteria %d",
```

Comment:
nit: move this statement up into its own `if len(allCFJobs) >= mcf` condition to make code more readable.